File: 101.ClosestPairOfPointsAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio       */
/* -----------------------------------------------------------------------  */

import java.io.*;
import java.util.*;

public class ClosestPairOfPoints {

    // Function to calculate Euclidean distance between two points
    static double dist(int[] p1, int[] p2) {
        // Formula: sqrt((x2 - x1)^2 + (y2 - y1)^2)
        return Math.sqrt(Math.pow(p1[0] - p2[0], 2) + Math.pow(p1[1] - p2[1], 2));
    }

    // Brute force method for smaller subsets (≤ 3 points)
    static double bruteForce(int[][] pts, int n) {
        double min = Double.MAX_VALUE;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                double d = dist(pts[i], pts[j]);
                if (d < min)
                    min = d;  // keep track of smallest distance found
            }
        }
        return min;
    }

    // Function to find the minimum distance within a strip region
    static double stripClosest(List<int[]> strip, double d) {
        double min = d;
        // Sort points by y-coordinate to compare only nearby points
        strip.sort(Comparator.comparingInt(p -> p[1]));

        for (int i = 0; i < strip.size(); i++) {
            int j = i + 1;
            // A point needs to be compared with next 7 points (geometry limit)
            while (j < strip.size() && (strip.get(j)[1] - strip.get(i)[1]) < min) {
                min = Math.min(min, dist(strip.get(i), strip.get(j)));
                j++;
            }
        }
        return min;
    }

    // Recursive function implementing Divide & Conquer
    static double closestUtil(int[][] pts, int n) {
        // Base case: when few points remain, use brute force
        if (n <= 3) return bruteForce(pts, n);

        int mid = n / 2;
        int[] midPoint = pts[mid]; // Midpoint for dividing

        // Recursively compute minimum distances in left & right halves
        double dl = closestUtil(Arrays.copyOfRange(pts, 0, mid), mid);
        double dr = closestUtil(Arrays.copyOfRange(pts, mid, n), n - mid);

        // Find smaller of the two distances
        double d = Math.min(dl, dr);

        // Build a strip containing points close (|x - mid_x| < d)
        List<int[]> strip = new ArrayList<>();
        for (int[] p : pts) {
            if (Math.abs(p[0] - midPoint[0]) < d)
                strip.add(p);
        }

        // Combine: check if strip gives smaller distance
        return Math.min(d, stripClosest(strip, d));
    }

    // Main function to find smallest distance
    static double closestPair(int[][] pts) {
        // Sort points initially by x-coordinate before recursion
        Arrays.sort(pts, Comparator.comparingInt(p -> p[0]));
        return closestUtil(pts, pts.length);
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // Example Input (Can be replaced with user input)
        int[][] pts = {
            {2, 3},
            {12, 30},
            {40, 50},
            {5, 1},
            {12, 10},
            {3, 4}
        };

        double res = closestPair(pts);
        System.out.println("The smallest distance is: " + res);
    }
}

File: 101.ClosestPairOfPointsAlgo.py
# The Closest Pair of Points problem finds two points in a 2D plane that are 
# closest to each other based on their Euclidean distance.
# It’s an important computational geometry problem solved 
# efficiently using the Divide and Conquer technique.

# Explanation

# A naïve approach compares every pair of points — taking O(n²) time.
# But with Divide and Conquer, we can reduce the complexity to O(n log n).

# Steps:

# Sort the points based on their x-coordinates.

# Divide the set of points into two halves (left and right).

# Recursively find the smallest distance in each half.

# Combine the results — check for closer pairs that 
# lie across the dividing line using a “strip” region.

# Return the minimum of all distances found.

import math

def dist(p1, p2):
    return math.sqrt((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2)

def bruteforce(points):
    n = len(points)
    min_d = float('inf')
    for i in range(n):
        for j in range(i + 1, n):
            d = dist(points[i], points[j])
            if d < min_d:
                min_d = d
    return min_d

def stripClosest(strip, d):
    min_d = d
    strip.sort(key = lambda p: p[1])

    for i in range(len(strip)):
        j = i + 1
        while j < len(strip) and (strip[j][1] - strip[i][1]) < min_d:
            min_d = min(min_d, dist(strip[i], strip[j]))
            j += 1
    return min_d

def closestUtil(points):
    n = len(points)
    if n <= 3:
        return bruteforce(points)
    
    mid = n // 2
    midPoint = points[mid]

    dl = closestUtil(points[:mid])
    dr = closestUtil(points[mid:])
    d = min(dl, dr)

    strip = [p for p in points if abs(p[0] - midPoint[0]) < d]

    return min(d, stripClosest(strip, d))

def closestPair(points):
    points.sort(key = lambda p: p[0])
    return closestUtil(points)

points = [(2, 3), (12, 30), (40, 50), (5, 1), (12, 10), (3, 4)]
print("The smallest distance is:", closestPair(points))





